pkg/controllers/dashboardapi/settings/settings_test.go

```diff
@@ -0,0 +1,129 @@
+package settings
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/stretchr/testify/assert"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type testCase struct {
+	description     string
+	envVar          *string
+	newSetting      settings.Setting
+	existingSetting *v3.Setting
+}
+
+var settingsETCD = make(map[string]*v3.Setting)
+
+// TestSetAll ensures that SetAll properly updates settings in ETCD. These updates
+// are informed by env vars of the form CATTLE_<SETTING_NAME> and the passed map
+// of settings.Setting objects.
+func TestSetAll(t *testing.T) {
+	settingClient := NewMockSettingClient(gomock.NewController(t))
+	testSettingsProvider := settingsProvider{
+		settings: settingClient,
+	}
+
+	testCases := populateTestCases()
+	settingClient.EXPECT().Get(gomock.Any(), gomock.Any()).DoAndReturn(func(name string, options metav1.GetOptions) (*v3.Setting, error) {
+		val := settingsETCD[name]
+		if val == nil {
+			return nil, errors.NewNotFound(schema.GroupResource{}, name)
+		}
+		return val, nil
+	}).AnyTimes()
+
+	settingClient.EXPECT().Create(gomock.Any()).DoAndReturn(func(setting *v3.Setting) (*v3.Setting, error) {
+		settingsETCD[setting.Name] = setting
+		return setting, nil
+	}).AnyTimes()
+
+	settingClient.EXPECT().Update(gomock.Any()).DoAndReturn(func(s *v3.Setting) (*v3.Setting, error) {
+		settingsETCD[s.Name] = s
+		return s, nil
+	}).AnyTimes()
+
+	settingMap := make(map[string]settings.Setting)
+	for _, testCase := range testCases {
+		settingMap[testCase.newSetting.Name] = testCase.newSetting
+		if testCase.envVar != nil {
+			envKey := settings.GetEnvKey(testCase.newSetting.Name)
+			os.Setenv(envKey, *testCase.envVar)
+			defer os.Unsetenv(envKey)
+		}
+
+		settingsETCD[testCase.newSetting.Name] = testCase.existingSetting
+	}
+	testSettingsProvider.SetAll(settingMap)
+	for _, testCase := range testCases {
+		finalSetting := settingsETCD[testCase.newSetting.Name]
```

Comment:
Nit: There doesn't appear to be any checks on the setting of testSettingsProvider.fallback. That's ok since that value is internal to the object, but I would recommend considering if we want to assert on the behavior of that value.

Reply:
Added checks for this